### PR TITLE
Writing stories in plain English, with words convertible to boolean

### PR DESCRIPTION
**Pull request by alb-i986**
BooleanWord is an enum that allows story writers to write natural prose, while developers can enjoy getting booleans "for free".

Quoting from the javadoc..

Example of use. Given the following candidate step:

```
 @Given("a stock of symbol $symbol $withThreshold a threshold")
 public void aStock(String symbol, BooleanWord withThreshold) {
     stock = new Stock(symbol).withThreshold(withThreshold.toBoolean());
 }
```

.. it is then possible to write the following scenarios:

```
 Scenario: a stock with threshold

 Given a stock of symbol AAA with a threshold
 When ..
 Then ..

 Scenario: a stock with no threshold

 Given a stock of symbol AAA without a threshold
 When ..
 Then ..
```

Please note that this depends on #77.

It would be nice if we could find an elegant way to internationalise the enum constants.
One where there is only one enum constant for each boolean word, and the translations are kept in an external file. Any ideas?

### CODE REVIEW

**Comment by maurotalevi** [comment]
Hi Alberto,

can you please raise JIRA issues for any contributions?

http://jbehave.org/how-to-contribute.html

Cheers

**Comment by alb-i986** [comment]
Hi Mauro,
sure, sorry about that.
I'll open the jiras and re-push with the Jira number in the commit msgs.
Thanks

**Comment by maurotalevi** [comment]
Cool - I think one JIRA would suffice for these commits - they are all 
in the same behavioural space.

**Comment by alb-i986** [comment]
> Cool - I think one JIRA would suffice for these commits - they are all in the same behavioural space.

I saw your comment after opening the two different jiras.
For now I've cleaned the two PRs and kept them separate.
Let me know if you still want me to join them.

Cheers

**Comment by maurotalevi** [comment]
It's fine - no probs.   I've pulled the first but I'm trying to 
understand how the second works.

Can you please provide more testing or examples of its use?

**Comment by alb-i986** [comment]
> Can you please provide more testing or examples of its use?

Maybe I could start from a clear problem statement :)
The problem is catching booleans from scenarios prose.

In my experience, typical use cases are:
- Given steps which say something like "Given some object *with/without* some property" (example which I reported in the javadoc)
- Then steps which say "Then I *should/should not* be able to do something"

I think two naive solutions may be:
- defining two different step methods, one matching "should", one matching "should not". But then they will probably be very very similar (=> duplication)
- defining the step method to take a boolean param. But then stories would contain "true"/"false" in the prose, making it less natural

With BooleanWord, developers can catch the booleans by just defining a parameter of type BooleanWord. All of the other stakeholders can read/write natural prose.

E.g.
https://skillsmatter.com/skillscasts/1087-a-year-of-misbehaving-a-retrospective-on-using-bdd-in-a-commercial-enterprise
At 11:55 I think the step at the bottom on the slide says "Then trader should <beOrNot> alerted"
In this case, I would write instead:
"Then trader <shouldBeAlerted> alerted"
And the step candidate:
@Then("trader <shouldBeAlerted> alerted")
public void traderShouldOrNotBeAlerted(BooleanWord shouldBeAlerted);

PS: When I watched the video I was wondering how you implemented that "<beOrNot>".

**Comment by maurotalevi** [comment]
I think the use case for a boolean value converter is valid one.

Perhaps the enum may not be the optimal route or implementation - 
amongst other things for the lack of i18n support you mention.

What about we define a BooleanValue object that takes a String value 
argument and looks up the true/false mapping?   This could support i18n  
too.

**Comment by alb-i986** [comment]
Ya, we should drop the enum implementation, and add a custom converter instead.
I think there's no need for a custom BooleanValue class. Boolean should do what we need.

I see it as follows:

1. Story file: `Given ... *with* ...`
2. Custom converter BooleanWordConverter: translates (if needed) and converts the value to Boolean
3. Candidate step with Boolean param: `public void givenAnObject(Boolean withProperty)`

BooleanWordConverter does the following:
1. takes the String value, i.e. `con` (Italian for "with")
2. translates it to English by looking up a property file, `boolean_words_it.properties` (similar to keywords_it.properties)
3. converts the English word to Boolean, by looking up another properties file, `boolean_word.properties`

```
$ cat boolean_words_it.properties
con=with
senza=without
```

```
$ cat boolean_words.properties
with=true
without=false
should=true
should not=false
```

**Comment by maurotalevi** [comment]
We should try to enhance existing Boolean converter to support file word lookup, rather than adding new converter.  The end converted object is the same.

Let's keep issue of i18n of parameter values separate as it can be relevant to other uses, as recently reported on lists.

**Comment by alb-i986** [comment]
I've worked on your last comment.
I enhanced BooleanConverter to convert any expression contained in a properties file.
It can also handle multiple locales.

```
String value = "with";
ParameterConverter converter = new BooleanConverter();
Boolean convertedValue = (Boolean) converter.convertValue(value, Boolean.class);
assertThat(convertedValue, is(true));
```

A few details I'd like to hear your opinion about:
- `withLocale` re-throws IOException as a ParameterConvertionFailed in case the loading of the properties file fails. Or should we just throw IOException?
- by default, BooleanConverter uses the English locale

**Comment by alb-i986** [comment]
If this was closed for the same reason as #78, there must be something wrong: here there is a [jira open](https://jbehave.atlassian.net/browse/JBEHAVE-1074), and the commit messages have the jira number

**Comment by maurotalevi** [comment]
Yes, it was closed by Github because of the branch rename.

Now reopened.    Please comment on 
https://jbehave.atlassian.net/browse/JBEHAVE-1074.
